xarray/ufuncs.py

```diff
@@ -0,0 +1,366 @@
+
+    return namespaces
+
+
+def get_array_namespace(*args):
+    xps = set()
+    for arg in args:
+        _walk_array_namespaces(arg, xps)
+
+    xps.discard(np)
+    if len(xps) > 1:
+        names = [module.__name__ for module in xps]
+        raise ValueError(f"Mixed array types {names} are not supported.")
+
+    return next(iter(xps)) if len(xps) else np
+
+
+class _UnaryUfunc:
+    """Wrapper for dispatching unary ufuncs."""
+
+    def __init__(self, name):
+        self._name = name
+
+    def __call__(self, x, **kwargs):
+        xp = get_array_namespace(x)
+        func = getattr(xp, self._name)
+        return xr.apply_ufunc(func, x, dask="allowed", **kwargs)
+
+
+class _BinaryUfunc:
+    """Wrapper for dispatching binary ufuncs."""
+
+    def __init__(self, name):
+        self._name = name
+
+    def __call__(self, x, y, **kwargs):
+        xp = get_array_namespace(x, y)
+        func = getattr(xp, self._name)
+        return xr.apply_ufunc(func, x, y, dask="allowed", **kwargs)
+
+
+class _UnavailableUfunc:
+    """Wrapper for unimplemented ufuncs in older numpy versions."""
+
+    def __init__(self, name):
+        self._name = name
+
+    def __call__(self, *args, **kwargs):
+        raise NotImplementedError(
+            f"Ufunc {self._name} is not available in numpy {np.__version__}."
+        )
+
+
+def _skip_signature(doc, name):
+    if not isinstance(doc, str):
+        return doc
+
+    if doc.startswith(name):
+        signature_end = doc.find("\n\n")
+        doc = doc[signature_end + 2 :]
+
+    return doc
+
+
+def _remove_unused_reference_labels(doc):
+    if not isinstance(doc, str):
+        return doc
+
+    max_references = 5
+    for num in range(max_references):
+        label = f".. [{num}]"
+        reference = f"[{num}]_"
+        index = f"{num}.    "
+
+        if label not in doc or reference in doc:
+            continue
+
+        doc = doc.replace(label, index)
+
+    return doc
+
+
+def _dedent(doc):
+    if not isinstance(doc, str):
+        return doc
+
+    return textwrap.dedent(doc)
+
+
+def _create_op(name):
+    if not hasattr(np, name):
+        # handle older numpy versions with missing array api standard aliases
+        if np.lib.NumpyVersion(np.__version__) < "2.0.0":
+            return _UnavailableUfunc(name)
+        raise ValueError(f"'{name}' is not a valid numpy function")
+
+    np_func = getattr(np, name)
+    if hasattr(np_func, "nin") and np_func.nin == 2:
+        func = _BinaryUfunc(name)
+    else:
+        func = _UnaryUfunc(name)
+
+    func.__name__ = name
+    doc = getattr(np, name).__doc__
+
+    doc = _remove_unused_reference_labels(_skip_signature(_dedent(doc), name))
+
+    func.__doc__ = (
+        f"xarray specific variant of numpy.{name}. Handles "
+        "xarray objects by dispatching to the appropriate "
+        "function for the underlying array type.\n\n"
+        f"Documentation from numpy:\n\n{doc}"
+    )
+    return func
```

Comment:
Type checkers can't evaluate code at runtime, so you really want to write something that prescribes the type signature statically, like:
```python
abs = _UnaryUfunc('abs')
```
or
```python
def _unary_ufunc(name: str) -> _UnaryUfunc:
    func = _UnaryUfunc(name)
    func.__doc__ = ...
    return func

abs = _unary_ufunc('abs')
```

When you write `abs = _create_op('abs')`, type checkers think the type of `abs` could be any of `_UnavailableUfunc` or `_BinaryUfunc` or `_UnaryUfunc`. (You can check this with `reveal_type(abs)` if you're curious.)